tests/test-update-direct.c

```diff
@@ -0,0 +1,255 @@
+  g_autoptr(GMainLoop) loop = g_main_loop_new (NULL, FALSE);
+  DownloadSource main_source = DOWNLOAD_MAIN;
+  gboolean has_commit;
+
+  /* We could get OSTree working by setting OSTREE_BOOTID, but shortly
+   * afterwards we hit unsupported syscalls in qemu-user when running in an
+   * ARM chroot (for example), so just bail. */
+  if (!eos_test_has_ostree_boot_id ())
+    {
+      g_test_skip ("OSTree will not work without a boot ID");
+      return;
+    }
+
+  server_root = g_file_get_child (fixture->tmpdir, "main");
+  server = eos_test_server_new_quick (server_root,
+                                      default_vendor,
+                                      default_product,
+                                      default_collection_ref,
+                                      0,
+                                      fixture->gpg_home,
+                                      keyid,
+                                      default_ostree_path,
+                                      NULL, NULL, NULL,
+                                      &error);
+  g_assert_no_error (error);
+  g_assert_cmpuint (server->subservers->len, ==, 1u);
+
+  subserver = g_object_ref (EOS_TEST_SUBSERVER (g_ptr_array_index (server->subservers, 0)));
+  client_root = g_file_get_child (fixture->tmpdir, "client");
+  client = eos_test_client_new (client_root,
+                                default_remote_name,
+                                subserver,
+                                default_collection_ref,
+                                default_vendor,
+                                default_product,
+                                &error);
+  g_assert_no_error (error);
+
+  g_hash_table_insert (subserver->ref_to_commit,
+                       ostree_collection_ref_dup (default_collection_ref),
+                       GUINT_TO_POINTER (1));
+  eos_test_subserver_update (subserver,
+                             &error);
+  g_assert_no_error (error);
+
+  eos_test_client_run_updater (client,
+                               &main_source,
+                               1,
+                               NULL,
+                               NULL,
+                               &error);
+  g_assert_no_error (error);
+
+  updater = eos_updater_proxy_new_for_bus_sync (G_BUS_TYPE_SESSION,
+                                                G_DBUS_PROXY_FLAGS_NONE,
+                                                "com.endlessm.Updater",
+                                                "/com/endlessm/Updater",
+                                                NULL,
+                                                &error);
+  g_assert_no_error (error);
+
+  g_signal_connect (updater, "notify::state",
+                    G_CALLBACK (updater_state_changed_cb),
+                    loop);
+
+  /* start the state changes */
+  updater_state_changed_cb (updater, NULL, loop);
+
+  g_main_loop_run (loop);
+
+  eos_test_client_has_commit (client,
```

Comment:
This only tests that the client has the given commit, not that it’s the currently deployed commit. I think you should also add something along the lines of `eos_test_client_has_commit (…, 2, …); g_assert_no_error (…); g_assert_false (has_commit);` to check that the update didn’t succeed at all.

Reply:
(I might’ve misunderstood what’s going on here, though, so correct me if I’m wrong.)

Reply:
I took most of this messy code from `test-update-from-main` which I believe should be testing for the deployed commit.
Looking into `utils.c`, the `eos_test_client_has_commit` starts by calling:
```
  if (!get_deploy_ids (sysroot, osname, &ids, error))
    return FALSE;
```
Before looking for the commit filename for the commit number. So I'd say it's looking for the right deployed commit here (but I've spent more time than I wanted looking at this code and still only partially understand what's going on such is the mess... so please bear with me :slightly_smiling_face: ).